Sources/MongoSwift/MongoError.swift

```diff
@@ -1,6 +1,137 @@
+public struct WriteError: Codable {
+    /// An integer value identifying the error.
+    public let code: ServerErrorCode
+
+    /// A description of the error.
+    public let message: String
+
+    private enum CodingKeys: String, CodingKey {
+        case code
+        case message = "errmsg"
+    }
+}
+
+/// A struct to represent a write concern error resulting from an executed bulk write.
+public struct WriteConcernError: Codable {
+    /// An integer value identifying the write concern error.
+    public let code: ServerErrorCode
+
+    /// A document identifying the write concern setting related to the error.
+    public let details: Document
+
+    /// A description of the error.
+    public let message: String
+
+    private enum CodingKeys: String, CodingKey {
+        case code
+        case details = "errInfo"
+        case message = "errmsg"
+    }
+}
+
+/// A struct to represent a write error resulting from an executed bulk write.
+public struct BulkWriteError: Codable {
+    /// An integer value identifying the error.
+    public let code: ServerErrorCode
+
+    /// A description of the error.
+    public let message: String
+
+    /// The index of the request that errored.
+    public let index: Int
+
+    /// The request that errored.
+    public let request: WriteModel? = nil
+
+    private enum CodingKeys: String, CodingKey {
+        case code
+        case message = "errmsg"
+        case index
+    }
+}
+
+/// Internal helper function used to get an appropriate error from a libmongoc error. This should NOT be used to get
+/// `.writeError`s or `.bulkWriteError`s. Instead, construct those from replies returned from libmongoc functions.
+internal func parseMongocError(error: bson_error_t, errorLabels: [String]? = nil) -> MongoSwiftError {
+    let domain = mongoc_error_domain_t(rawValue: error.domain)
+    let code = mongoc_error_code_t(rawValue: error.code)
+    let message = toErrorString(error)
+
+    switch (domain, code) {
+    case (MONGOC_ERROR_CLIENT, MONGOC_ERROR_CLIENT_AUTHENTICATE):
+        return RuntimeError.authenticationError(message: message)
+    case (MONGOC_ERROR_COMMAND, MONGOC_ERROR_COMMAND_INVALID_ARG):
+        return UserError.invalidArgumentError(message: message)
+    case (MONGOC_ERROR_SERVER, _):
+        return ServerError.commandError(
+                code: ServerErrorCode(code.rawValue),
+                message: message,
+                errorLabels: errorLabels)
+    case (MONGOC_ERROR_STREAM, _):
+        return RuntimeError.connectionError(message: message, errorLabels: errorLabels)
+    case (MONGOC_ERROR_SERVER_SELECTION, MONGOC_ERROR_SERVER_SELECTION_FAILURE):
+        return RuntimeError.connectionError(message: message, errorLabels: errorLabels)
+    case (MONGOC_ERROR_PROTOCOL, MONGOC_ERROR_PROTOCOL_BAD_WIRE_VERSION):
+        return RuntimeError.connectionError(message: message, errorLabels: errorLabels)
+    default:
```

Comment:
I've added the assert. I'd rather not expand the labels to the other types (non-server, non-connection) if they will probably always be nil. From my reading of the transactions spec, we should be relatively safe to keep them only on what we have so far, but I'm not sure how subject that is to change. Do you think there is a reasonable risk that error labels could become more pervasive?

Reply:
I think connection and server errors are sufficient. I'm hard-pressed to think of a case where an auth error might occur during a transaction (perhaps on a new socket being used for the first time), but even then I don't think an error label would be relevant since the driver would not have accomplished anything with the socket.

This looks fine as-is.